client/components/Dashboard.tsx

```diff
@@ -20,34 +17,153 @@ const GET_PROJECT_DATA = gql`
     }
 `;
 
+const UPDATE_RATE_LIMITER_CONFIG_MUTATION = gql`
+    mutation udpateRateLimiter($projectId: String!, $rateLimiterConfig: RateLimiterConfigInput) {
+        updateProject(id: $projectId, rateLimiterConfig: $rateLimiterConfig) {
+            rateLimiterConfig {
+                type
+                options {
+                    ... on WindowOptions {
+                        capacity
+                        windowSize
+                    }
+                    ... on BucketOptions {
+                        capacity
+                        refillRate
+                    }
+                }
+            }
+        }
+    }
+`;
+
+const GET_RATE_LIMITER_CONFIG_QUERY = gql`
+    query getRateLimiter($projectId: String!) {
+        project(id: $projectId) {
+            rateLimiterConfig {
+                type
+                options {
+                    ... on WindowOptions {
+                        capacity
+                        windowSize
+                    }
+                    ... on BucketOptions {
+                        capacity
+                        refillRate
+                    }
+                }
+            }
+        }
+    }
+`;
+
+interface ProjectResult {
+    project: Project;
+}
+interface ProjectVars {
+    projectId: string;
+}
+
+interface UpdateRateLimiterVars extends ProjectVars {
+    rateLimiterConfig: RateLimiterConfig;
+}
+
 export default function Dashboard() {
     /** Bring the user context into this component */
     const { user } = useAuth();
 
-    /** State requirments for this component */
+    /** State requirements for this component */
     const [selectedProject, setSelectedProject] = useState<Project>();
-    const [allProjects, setAllProjects] = useState<Project[]>();
 
+    // Apollo graphql hooks
     /** Send query to get project information for this user */
-    const { data, loading } = useQuery(GET_PROJECT_DATA, {
-        variables: {
-            userId: user!.id,
-        },
-    });
+    const [getUserData, userData] = useLazyQuery<{ user: User }, { userId: string }>(
+        GET_PROJECT_DATA
+    );
+
+    const [getRateLimiterConfig, rateLimitResponse] = useLazyQuery<ProjectResult, ProjectVars>(
+        GET_RATE_LIMITER_CONFIG_QUERY,
+        { fetchPolicy: 'network-only' }
+    );
+
+    const [udpateRateLimiter] = useMutation<ProjectResult, UpdateRateLimiterVars>(
+        UPDATE_RATE_LIMITER_CONFIG_MUTATION
+    );
+
     useEffect(() => {
-        if (!loading && data) {
-            setAllProjects(data.user.projects);
+        if (user?.id) {
+            getUserData({
+                variables: {
+                    userId: user.id,
+                },
+            });
         }
-    }, [loading, data]);
+    }, [user]);
+
+    useEffect(() => {
+        // Fetches Rate Limiter settings whenever project is changed
+        if (selectedProject?.id) {
+            (async () => {
```

Comment:
Is there an advantage to this being a self invoking async function as opposed to the useEffect being async?

Reply:
Passing an async callback to useEffect is discouraged. useEffect expects a function that returns unedefined or a callback. Not a promise.